Refactor http2_grpc to leverage tail calls

rafaelroquetto opened the pull request:
The latest kernel versions have changed the way the verifier deals with scalar spills, reducing its tolerance towards complex programs - for a comprehensive discussion, please refer to

  https://lore.kernel.org/bpf/50b7301fcfd0682c9923b3639c1589f3eb37af33.camel@crowdstrike.com/T/

Our http2 code therefore no longer passes the verifier, owing to mostly a complex _for_ loop. This PR splits the code into different bpf programs that are now tail called, and changes the `_for_` loop to be a hybrid of a recursive bpf_tail_call and a loop with less iterations. This approach was chosen because neither do we want to bpf_tail_call/recurse for each iteration, nor do we wish to use `bpf_loop` as constrains our program to kernel versions greater than 5.17.

rafaelroquetto commented on bpf/k_tracer_tailcall.h:
```diff
@@ -0,0 +1,22 @@
+#ifndef K_TRACER_TAILCALL_H
```
This file was necessary to avoid a cyclic header dependency.

rafaelroquetto commented on bpf/protocol_http2.h:
```diff
@@ -24,6 +25,30 @@ struct {
+    http2_grpc_request_t prev_info;
+    u8 has_prev_info;
+
+    int pos; //FIXME should be size_t equivalent
```
This FIXME will be addressed separately - there are too many other places to check and tweak.

rafaelroquetto replied:
(copying this here as I commented on the wrong box - sorry about the noise)

With regards to the int usage, my biggest concern is that we have a signed int here - in most APIs (except from a few relics like libc functions such as read(), which will rely on negative values for errors, and anything else for "byte count"), we will encounter size_t, which tends to be the same as uintptr_t because it needs to be able to address the memory space. We do invariably end up mixing these ints up with pointers and alike, which are unsigned by default, so IMHO it's just best to eventually refactor them to become unsigned (and simplify underflow/overflow cases).

Using a smaller unsigned int here is potentially fine, but we have to be careful to not make it appear as an lvalue in an operation involving pointers as rvalues, as it can cause an overflow (because pointers will be 64-bit). It becomes a trade-off between saving a few bytes (or a lot depending on how many of those we have) versus less cognitive load of treating everything as size_t.

mariomac reviewed (commented):
LGTM as long as test are fixed!

Also, could you add some extra commenting to the new C functions 🙏 ?

grcevski commented on bpf/protocol_http2.h:
```diff
@@ -95,110 +125,219 @@ http2_grpc_end(http2_conn_stream_t *stream, http2_grpc_request_t *prev_info, voi
+        bpf_map_lookup_elem(&ongoing_http2_grpc, &g_ctx->stream);
+
+    if (prev_info) {
+        g_ctx->prev_info = *prev_info;
```
I like this change, it's safer to make a copy of the prev info rather than to keep a pointer. With high transaction rates we can potentially pick up stale data.

rafaelroquetto replied:
I was also having trouble keeping the pointer on a map. If I stored it as a `u64` and tried to cast it back and deference it later on, the verifier would complain about "forbidden memory deref" or something like that.

grcevski commented on bpf/protocol_http2.h:
```diff
@@ -95,110 +125,219 @@ http2_grpc_end(http2_conn_stream_t *stream, http2_grpc_request_t *prev_info, voi
+    // case we try this backup path, which will tail call on success.
+    handle_data_frame(ctx, g_ctx);
+
+    // we found nothing, tough luck!
```
Heh, I like the comment but it's also not unusual to find nothing :). HTTP2/gRPC will exchange many frames that are related to the data being sent around, so it's very normal not to find anything except when the transaction starts and ends.

rafaelroquetto replied:
Sheer luck then! :) I will nuke that comment.

grcevski reviewed (commented):
Great refactor, I left a few comments and nits. I think the only one that must be done is removing the prints in the loop and I think it's important to address my question on returning frame_header_t.

grcevski commented:
The test failure in integration tests looks related to the change. I can help debug if you need help :)

rafaelroquetto commented on bpf/protocol_http2.h:
```diff
@@ -95,110 +125,219 @@ http2_grpc_end(http2_conn_stream_t *stream, http2_grpc_request_t *prev_info, voi
+        return 0;
+    }
+
+    for (int i = 0; i < 4; i++) {
```
Good idea, will do!

rafaelroquetto commented on bpf/http2_grpc.h:
```diff
@@ -8,8 +8,6 @@
 
-#define FRAME_HEADER_LEN 9
-
 #define FLAG_DATA_END_STREAM 0x1
```
Agreed, I missed that

rafaelroquetto commented on bpf/protocol_http2.h:
```diff
@@ -95,110 +125,219 @@ http2_grpc_end(http2_conn_stream_t *stream, http2_grpc_request_t *prev_info, voi
+// TAIL_PROTOCOL_HTTP2_GRPC_HANDLE_START_FRAME
+SEC("kprobe/http2")
+int protocol_http2_grpc_handle_start_frame(void *ctx) {
+    (void)ctx;
```
Silences `-Wunused-parameter`. Alternatively one could omit the parameter name but that's a C23 extension.

rafaelroquetto commented on bpf/protocol_http2.h:
```diff
@@ -95,110 +125,219 @@ http2_grpc_end(http2_conn_stream_t *stream, http2_grpc_request_t *prev_info, voi
-
-    for (int i = 0; i < 8; i++) {
-        if (pos >= bytes_len) {
+static __always_inline frame_header_t next_frame(const grpc_frames_ctx_t *g_ctx) {
```
The compiler will always optimise the return value here, even with -O0 - both gcc and clang will do it. It may not optimise with `-O0`  case for C (it will for C++ >=17 because of [mandatory copy elision)](https://en.cppreference.com/w/cpp/language/copy_elision) for the general case, but it has to and will for our particular case because of `__always_inline` - as a matter of fact, it won't even compile otherwise, as bpf does not support returning aggregates. Check out [this example](https://godbolt.org/z/4nYe8WY4r). So that guarantees it for us.

Passing things as out parameters should be avoided when possible because it prevents [const-correctness](https://isocpp.org/wiki/faq/const-correctness).

I would still like to keep it as is for clarity and the reasons above (const is something we should strive to use more, as it will shift a few bugs to compile time and increase clarity), even if it costs a few instructions - which actually will also not be the case. For no optimisation cases, the compiler will allocate the return value on the stack and then read from it - for optimised cases, the compiler will read directly from the return register if we return a `u8`, but will have to read from a stack offset in the present case, which is a negligible penalty, if anything (it's probably hot on cache anyway), which is offset in case we need to pass a in-out pointer anyway.

rafaelroquetto replied:
_Update:_ I ended up removing the return value (and the extra enum) altogether as they were causing the loop to bail earlier and causing a bug. The correct behaviour, consistent with the pre-refactor code, is to continue looping into the next frame.

grcevski replied:
Ok, great. I really think we should avoid relying on clang will optimize this away in a form which we can easily write.

rafaelroquetto commented on bpf/protocol_http2.h:
```diff
@@ -95,110 +125,219 @@ http2_grpc_end(http2_conn_stream_t *stream, http2_grpc_request_t *prev_info, voi
+        goto header_error;
+    }
+
+    bpf_dbg_printk("http2 frame type = %d, len = %d", header.type, header.length);
```
With regards to the `int` usage, my biggest concern is that we have a signed int here - in most APIs (except from  a few relics like libc functions such as read(), which will rely on negative values for errors, and anything else for "byte count"), we will encounter `size_t`, which tends to be the same as `uintptr_t` because it needs to be able to address the memory space. We do invariably end up mixing these ints up with pointers and alike, which are unsigned by default, so IMHO it's just best to eventually refactor them to become unsigned (and simplify underflow/overflow cases).

Using a smaller _unsigned_ int here is potentially fine, but we have to be careful to not make it appear as an lvalue in an operation involving pointers as rvalues, as it can cause an overflow (because pointers will be 64-bit). It becomes a trade-off between saving a few bytes (or a lot depending on how many of those we have) versus less cognitive load of treating everything as `size_t`.

rafaelroquetto replied:
Done

rafaelroquetto commented:
Tests now fixed - caused by a typo (missing `!` in `if (!frame.length ...)` now reworded as `if (frame.length == 0)` to make it more obvious and by the introduction of `FrameUknown`, which caused the loop to bail earlier than it should (and not parse incoming frames), which I have now removed, making it consistent with the old code.

Apart from that, I've incorporated most review feedbacks.

grcevski reviewed (approved):
LGTM! Good stuff!
